### PR TITLE
Improve pid_is_alive(): Check if the pid belongs to darktable

### PR DESCRIPTION
**Pull request by christte**
When restarting darktable after a crash, the database will be locked. A check is performed to see if darktable is still running, using pid_is_alive().

The problem is that pid_is_alive() does not verify that the pid noted in the database lock is actually a darktable instance.

In my case I could not launch darktable after a crash, because the previous pid was taken by another process. This patch fixes that problem.

### CODE REVIEW

File: src/common/database.c [hunk 1]
```diff
@@ -1371,6 +1371,30 @@ static gboolean pid_is_alive(int pid)
   }
 #else
   pid_is_alive = !((kill(pid, 0) == -1) && errno == ESRCH);
+
+#ifdef __linux__
+  // If this is Linux, we can query /proc to see if the pid is
+  // actually a darktable instance.
+  if (pid_is_alive)
+  {
+    char filename[64];
```
Comment (LebedevRI) [inline src/common/database.c]:
```c
char filename[sizeof(pid)+1];
``` 
?

File: src/common/database.c [hunk 2]
```diff
@@ -1371,6 +1371,30 @@ static gboolean pid_is_alive(int pid)
...
+    char cmdline[MAX_PATH+1];
```
Comment (LebedevRI) [inline src/common/database.c]: `char cmdline[PATH_MAX];`

File: src/common/database.c (hunk 1 above)
Comment (LebedevRI) [inline src/common/database.c]: What is the max value of pid in linux, 32bit?

File: src/common/database.c [hunk 3]
```diff
@@ -1371,6 +1371,30 @@ static gboolean pid_is_alive(int pid)
...
+    sprintf(filename, "/proc/%d/cmdline", pid);
```
Comment (LebedevRI) [inline src/common/database.c]: `snprintf(filename, sizeof(filename), "/proc/%d/cmdline", pid);`

File: src/common/database.c [hunk 4]
```diff
@@ -1371,6 +1371,30 @@ static gboolean pid_is_alive(int pid)
...
+    FILE* f = fopen(filename, "r");
+    if (f) {
+      size_t size;
+      size = fread(cmdline, sizeof(char), MAX_PATH, f);
```
Comment (LebedevRI) [inline src/common/database.c]: `size = fread(cmdline, sizeof(char), sizeof(cmdline), f);`

File: src/common/database.c [hunk 5]
```diff
@@ -1371,6 +1371,30 @@ static gboolean pid_is_alive(int pid)
...
+      if (size > 0)
+      {
+        if (strstr(cmdline, "darktable") == NULL) {
```
Comment (LebedevRI) [inline src/common/database.c]: That will work as long the original dt executable contained that case-sensitive sub-string in the full path to the `argv[0]`.
I'm wondering if that is a safe assumption? Seems like it?

**Comment by christte** [comment]
Going to fix this up a bit + get in support for osx.

**Comment by houz** [comment]
I didn't add such a check originally as I am not convinced that we can rely on this working. However, in almost all cases it's probably fine, so why the heck not.
One general remark: Please fix your formatting, add no space after `if` and put `{` in their own line.

File: src/common/database.c (hunk 1 above)
Comment (christte) [inline src/common/database.c]: pid_t is 32bit int.

64 should be sufficient: strlen("/proc/[int64_max]/cmdline") < 64.

File: src/common/database.c (hunk 5 above)
Comment (christte) [inline src/common/database.c]: This should always work as long as the name of the dt binary contains 'darktable'. This also works when dt is launched from a script.

Comment (houz) [inline src/common/database.c]: The problem I see is having some other script inside the darktable folder which is running with that PID. It would trigger the test ("darktable" is in the directory name). However, that is such a fringe corner case that we can ignore it.

If we wanted to do the right thing we would have to check /proc/<pid>/fd/* and see if any one of those is pointing to the sqlite file. That's just a remark, please don't add code like that. :-)
